File: Python/N-Queens II.py
class Solution:
    # @return an integer
    def generate(self, pos, n, vis, res):
        if pos == n:
            res += 1
            return res
        for i in range(n):
            if not vis[0][i] and not vis[1][pos + i] and not vis[2][pos - i + n]:
                vis[0][i] = vis[1][pos + i] = vis[2][pos - i + n] = 1
                res = self.generate(pos+1, n, vis, res)
                vis[0][i] = vis[1][pos + i] = vis[2][pos - i + n] = 0
        return res
    def totalNQueens(self, n):
        vis = [[0 for col in range(30)] for row in range(3)]
        res = 0
        return self.generate(0, n, vis, res)

File: C++/N-Queens-II.cpp
class Solution {
public:
	void helper(vector<vector<int> >& vis, int pos, int n, int& res) {
        if (pos >= n) {
            res ++;
            return;
        }
        for (int i = 0; i < n; i ++) {
            if (vis[0][i] == 0 && vis[1][n + i - pos] == 0 && vis[2][i + pos] == 0) {
                vis[0][i] = vis[1][n + i - pos] = vis[2][i + pos] = 1;                
                helper(vis, pos + 1, n, res);
                vis[0][i] = vis[1][n + i - pos] = vis[2][i + pos] = 0;
            }
        }
    }
    int totalNQueens(int n) {
    	int res = 0;
        vector<vector<int> > vis(3, vector<int>(30, 0));
        helper(vis, 0, n, res);
        return res;
    }
};